### ISSUE TITLE
Supporting streaming in grpc-web

### ISSUE THREAD

**Issue by BatyrSeven** [stephenh/ts-proto#141]
I'm trying to compile the following proto file for grpc web:
```
service Storage {
  rpc upload (stream StorageUploadChunk) returns (StorageUploadResponse);
}

message StorageUploadChunk {
  bytes payload = 1;
}

message StorageUploadResponse {
  // some response data
}
```

with the following options for ts-proto:
```
env=browser
useOptionals=true
oneof=unions
outputClientImpl=grpc-web
configureProtobuf=true
```

However, the generated code has the error: 'Value of type 'typeof Observable' is not callable. Did you mean to include 'new'?'

The generated code:
```javascript
export class StorageClientImpl implements Storage {

  private readonly rpc: Rpc;

  constructor(rpc: Rpc) {
    this.rpc = rpc;
  }

  upload(request: DeepPartial<Observable<StorageUploadChunk>>, metadata?: grpc.Metadata): Promise<StorageUploadResponse> {
    return this.rpc.unary(StorageuploadDesc, Observable<StorageUploadChunk>.fromPartial(request), metadata);
  }

```

As you can see here, we are trying to call `fromPartial` function from `Observable`, not from `StorageUploadChunk`.

So the question is the following:
- Is it possible to use stream methods without Observable but with Promise? In my project,  the stream method is required by one feature only and I would be happy to know if there is any way to avoid connecting `rxjs` library to support Observable. 

P.S. It also can be noted that option `--ts_proto_opt=returnObservable=true` doesn't change the result.

**Comment by stephenh** [comment]
Yeah, streaming is not supporting for either requests or responses for grpc-web. I don't personally work with grpc-web or streaming, happy to help on a contracting/consulting basis, but otherwise PRs would be great. Thanks!

**Comment by ssilve1989** [comment]
`grpc-web` **does** support server streaming when using base64 encoding `grpcwebtext` but not binary encoding. See https://github.com/grpc/grpc-web#wire-format-mode

**Comment by stephenh** [comment]
@ssilve1989 ah sure, by my comment I meant "not supported _by ts-proto_", not necessarily within grpc-web itself; I'll update the comment to clarify. Thanks for the link!

**Comment by ssilve1989** [comment]
@stephenh Whoops 😅

**Comment by brianstarke** [comment]
I'm also interested in this seeing as I'd rather use this generator than the one we currently use. I'm considering implementing this in a fork and welcome comments, especially since I only have an idea about *returning* a stream - not sending one. I've no experience with sending one...

Given

```
service Foo {
  rpc bar (Baz) returns (stream Qux);
}

message Baz {
  string thing = 1;
}

message Qux {
  string current_status = 1;
}
```

I would generate something like

```ts
  Bar(req: Baz, onMessage: (msg: Qux) => void, onEnd: (error: Error) => void): void {
    const data = req.encode(req).finish();
    rpc.invoke(FooService.Bar, req, onMessage, onEnd);
    return;
  }
```

And the `rpc` interface would have an additional `invoke` method.

In `grpc-web` in particular, this implementation is just something like

```ts
grpc.invoke(Foo.Bar, {
      request: req,
      host: this._host,
      metadata: { auth: this._auth },
      onMessage: onMessage,
      onEnd: (code: grpc.Code, msg: string | undefined, trailers: grpc.Metadata) => {
        onEnd(new Error(`${grpc.Code[code]} - ${msg}`));
      },
    });
```

Thoughts? I can't think of a clean way to pass a `stream` just yet...

**Comment by brianstarke** [comment]
For a more concrete example I futzed around and did this in the `simple` integration test...

```
service PingService {
  rpc ping(PingRequest) returns(PingResponse);
  rpc pings(PingRequest) returns(stream PingResponse);
}
```

which generates

```ts
export interface PingService {

  ping(request: PingRequest): Promise<PingResponse>;

  pings(request: PingRequest, onMessage: (msg: PingRequest) , onEnd: (err: Error) ): void;

}
```

and 

```ts
export class PingServiceClientImpl implements PingService {

  private readonly rpc: Rpc;

  constructor(rpc: Rpc) {
    this.rpc = rpc;
  }

  ping(request: PingRequest): Promise<PingResponse> {
    const data = PingRequest.encode(request).finish();
    const promise = this.rpc.request("simple.PingService", "ping", data);
    return promise.then(data => PingResponse.decode(new Reader(data)));
  }

  pings(request: PingRequest, onMessage: (msg: PingResponse) , onEnd: (err: Error) ): void {
    const data = PingRequest.encode(request).finish();
    const _onMessage = data => onMessage(PingResponse.decode(new Reader(data)));
    this.rpc.invoke("simple.PingService", "pings", data, _onMessage, onEnd);
    return;
  }

}
```

**Comment by ngrok-sarid** [comment]
@stephenh Would it be possible to prevent the output of the code somehow until streams get added? i.e. from the first post:
```typescript
Observable<StorageUploadChunk>.fromPartial(request)
```

is not valid typescript/javascript code since `fromPartial` doesn't exist on `Observable`. This causes typing issues even if you never plan on using (from the example) `upload` from `StorageClientImpl`.

**Comment by timostamm** [comment]
> `grpc-web` **does** support server streaming (not client streaming like this issue is requesting) when using base64 encoding `grpcwebtext` but not binary encoding. See https://github.com/grpc/grpc-web#wire-format-mode

Going off-topic here, but grpc-web supports server streaming regardless of the encoding. 

Screenshot below shows several response messages and some trailers, all in binary encoding:

![Bildschirmfoto 2020-11-24 um 20 11 26](https://user-images.githubusercontent.com/4289451/100140623-8287b700-2e91-11eb-8ca4-1fcec342080a.png)


grpc-web is actually very simple. It uses simple length-delimited frames (DATA and TRAILER). The text-mode simply encodes every in base64.

**Comment by stephenh** [comment]
Fwiw I think @PhilipMantrov has this basically implemented in #131 using rxjs, which I think is a reasonable approach.

**Comment by PhilipMantrov** [comment]
> Fwiw I think @PhilipMantrov has this basically implemented in #131 using rxjs, which I think is a reasonable approach.

Yep. 
And I already have experiment with long pooling and bi-derectional streams. Later I will prepare pr. ^^

**Comment by BatyrSeven** [comment]
> Fwiw I think @PhilipMantrov has this basically implemented in #131 using rxjs, which I think is a reasonable approach.

Does this mean that steaming is now supported? I'm still getting the same error with the example in the first post (using ts-proto version 1.41.1)

**Comment by stephenh** [comment]
@BatyrSeven streaming responses are supported, streaming requests probably don't work yet. I forgot this issue was still open; going to close it as fixed, and if you need streaming request support, let's file a follow up issue for that.
